File: kattis/10kindsofpeople.cc
// https://open.kattis.com/problems/10kindsofpeople
#include <iostream>
#include <vector>

using namespace std;

typedef vector<int> vi;
typedef vector<bool> vb;
typedef vector<vb> vvb;

int find(vi &l, int x) {
	while (x != l[x]) x = l[x];
	return x;
}

void unite(vi &l, vi &s, int a, int b) {
	a = find(l, a);
	b = find(l, b);
	if (a == b) return;
	if (s[a] < s[b]) swap(a, b);
	s[a] += s[b];
	l[b] = a;
}

int main() {
	int r, c;
	cin >> r >> c;
	vi l(r * c), s(r * c, 1);
	vvb a(r);
	for (int i = 0; i < r; i++) {
		a[i] = vb(c, false);
		for (int j = 0; j < c; j++) {
			l[i * c + j] = i * c + j;
			char b;
			cin >> b;
			if (b == '1') a[i][j] = true;
		}
	}
	for (int i = 0; i < r; i++) {
		for (int j = 0; j < c; j++) {
			if (j < c - 1 && a[i][j] == a[i][j + 1])
				unite(l, s, i * c + j, i * c + j + 1);
			if (i < r - 1 && a[i][j] == a[i + 1][j])
				unite(l, s, i * c + j, (i + 1) * c + j);
		}
	}
	int n;
	cin >> n;
	while (n--) {
		int r1, c1, r2, c2;
		cin >> r1 >> c1 >> r2 >> c2;
		r1--; c1--; r2--; c2--;
		if (find(l, r1 * c + c1) == find(l, r2 * c + c2))
			cout << (a[r1][c1] ? "decimal\n" : "binary\n");
		else cout << "neither\n";
	}
}

File: kattis/10kindsofpeople.py
#!/usr/bin/env python2
# https://open.kattis.com/problems/10kindsofpeople
def find(l, x):
    while x != l[x]: x = l[x]
    return x

def unite(l, s, a, b):
    a = find(l, a)
    b = find(l, b)
    if a == b: return
    if s[a] < s[b]: a, b = b, a
    s[a] += s[b]
    l[b] = a

r, c = map(int, raw_input().split())
l = [i for i in range(r * c)]
s = [1] * r * c
a = [None] * r
for i in xrange(r):
    a[i] = map(int, raw_input())
for i in xrange(r):
    for j in xrange(c):
        if j < c - 1 and a[i][j] == a[i][j + 1]:
            unite(l, s, i * c + j, i * c + j + 1)
        if i < r - 1 and a[i][j] == a[i + 1][j]:
            unite(l, s, i * c + j, (i + 1) * c + j)
for _ in xrange(input()):
    r1, c1, r2, c2 = [int(x) - 1 for x in raw_input().split()]
    if find(l, r1 * c + c1) == find(l, r2 * c + c2):
        print 'decimal' if a[r1][c1] else 'binary'
    else: print 'neither'
